readme.md

js-ajax [![Build Status](https://travis-ci.org/tsu-complete/js-ajax.svg?branch=master)](https://travis-ci.org/tsu-complete/js-ajax) [![Dependency Status](https://david-dm.org/tsu-complete/js-ajax.svg)](https://david-dm.org/tsu-complete/js-ajax) [![Documentation Coverage](http://inch-ci.org/github/tsu-complete/js-ajax.svg?branch=master)](http://inch-ci.org/github/tsu-complete/js-ajax?branch=master)
===

> Just another javascript ajax utility

[![Browser Support](https://ci.testling.com/tsu-complete/js-ajax.png)
](https://ci.testling.com/tsu-complete/js-ajax)

License
---

[![WTFPL](http://www.wtfpl.net/wp-content/uploads/2012/12/wtfpl-badge-1.png)](http://www.wtfpl.net)

Usage
---

Promise wrapper

Options may be provided after or with the url

```js
ajax("path/to/service", { /* options */ });

// --or--

ajax({
    url:    "path/to/service",
    method: ajax.METHOD_GET,
    params: {
        "key", "value",
        "...", "..."
    }
});
```

Possible methods are

```
ajax.METHOD_GET
ajax.METHOD_POST
ajax.METHOD_PUT
ajax.METHOD_UPDATE
ajax.METHOD_DELETE
```

Returns a promise

```js
ajax("path/to/service").then(function ( data ) {
    // data code...
}).catch(function ( status ) {
    // error code...
});

// --or --

ajax("path/to/service").then(function ( data ) {
    // code...
}, function( status ) {
    // error code...
});
```

A request may be delayed
```js
var req = ajax("path/to/service", {auto:false});

// later...

req.send().then(...);
```

A sent request may be canceled

```js
ajax("path/to/service").cancel()
```

noConflict method supplied when used globally

```js
ajax = window.ajax.noConflict();
```

Linting
---

Linted with jshint

```js
$ npm run lint
```

Testing
---

Tested with mocha

```bash
$ npm run test
```

Documentation
---

Documented with jsdoc

```bash
$ npm run docs
```



changelog.md

v2.0.0
---
- moves to functional utility
- returns js promise for compatibility
- removes watch functionality

v2.0.1
--
- introduces content type

v2.0.2
---
- removes content length (unsafe)



test/setup.js

/* global chai */
(function ( ) {
    "use strict";

    if ("undefined" !== typeof window) {
        window.global = window;
    } else {
        global.window = global;
    }

    global.ENV_NODE = "node";
    global.ENV_BROWSER = "browser";

    if ("undefined" !== typeof process) {
        global.chai = require("chai");
        global.ENV = global.ENV_NODE;
    } else {
        global.ENV = global.ENV_BROWSER;
    }

    global.expect = chai.expect;
    global.assert = chai.assert;

})( );


test/ajax.test.js

/* global expect, ajax */
(function ( ) {
    "use strict";
    var URL, nock, server, id;

    URL = "http://localhost:3000";

    if (global.ENV === global.ENV_NODE) {
        global.ajax = require("../ajax");
        nock = require("nock");

        // xhr isnt going to exist at this point
        describe("missing support", function ( ) {
            it("should error without support", function ( ) {
                expect(function ( ) {
                    ajax(URL + "/respond/", {auto: false});
                }).to.throw(Error);
                // now polyfill!
                global.XMLHttpRequest = require("xhr2");
            });
        });

        // if we polyfill here support will exist before the first
        // test is run and the test will not throw the expected error

        // setup mock server
        id = (function ( ) {
            var _id = 0;
            return function ( ) { return _id++; };
        })( );

        before(function ( ) {
            server = nock(URL)
                .persist()

                .get("/respond/")
                .reply(200, "response")
                .get("/get_200/")
                .reply(200, "content")
                .get("/get_delay/")
                .delay(2000)
                .reply(200, "content")
                .get("/get_404/")
                .reply(404)
                .post("/post_200/", { a: 1, b: 2 })
                .reply(200, "content");
        });

        // clean up mock server
        after(function ( ) {
            nock.cleanAll();
        });
    }

    describe("environment", function ( ) {
        it("should have window", function ( ) {
            return expect(window).to.be.ok;
        });
        it("should have global", function ( ) {
            return expect(global).to.be.ok;
        });
        it("should have env", function ( ) {
            return expect(global.ENV).to.be.ok;
        });
    });

    describe("mock server", function ( ) {
        it("should respond", function ( done ) {
            ajax(URL + "/respond/").then(function ( data ) {
                data = expect(data).to.be.ok;
                done();
            }).catch(function ( status ) {
                throw new Error("unexpected status: " + status);
            });
        });
    });

    describe("argument normalization", function ( ) {
        it("should not accept no arguments", function ( ) {
            expect(function ( ) {
                ajax();
            }).to.throw(Error);
        });
        it("should accept string", function ( ) {
            expect(function ( ) {
                ajax(URL + "/get_200/");
            }).to.not.throw(Error);
        });
        it("should accept string, empty object", function ( ) {
            expect(function ( ) {
                ajax(URL + "/get_200/", { });
            }).to.not.throw(Error);
        });
        it("should not accept empty object", function ( ) {
            expect(function ( ) {
                ajax({ });
            }).to.throw(Error);
        });
        it("should accept object with url", function ( ) {
            expect(function ( ) {
                ajax({ url: URL + "/get_200/" });
            }).to.not.throw(Error);
        });
        it("should ignore unknown arugments", function ( ) {
            expect(function ( ) {
                ajax(URL + "/get_200/", { extra: 1, args: 2 });
            }).to.not.throw(Error);
        });
        it("should ignore extra arguments", function ( ) {
            expect(function ( ) {
                ajax(URL + "/get_200/", { }, 1, 2, 3);
            }).to.not.throw(Error);
        });
        it("should ignore non-string method", function ( ) {
            expect(function ( ) {
                ajax(URL + "/get_200/", { method: 1 });
            }).to.not.throw(Error);
        });
    });

    describe("callbacks", function ( ) {
        describe("where file exists", function ( ) {
            describe("then", function ( ) {
                it("should return content", function ( done ) {
                    ajax(URL + "/get_200/").then(function ( data ) {
                        expect(data).to.equal("content");
                        done();
                    });
                });
                it("should be cancelable", function ( done ) {
                    ajax(URL + "/get_delay/").cancel()
                    .then(function ( data ) {
                        throw new Error("unexpected data: " + data);
                    }, function ( status ) {
                        expect(status).to.equal(ajax.STATUS_ABORT);
                        done();
                    });
                });
                it("post should return content", function ( done ) {
                    ajax(URL + "/post_200/", {
                        method: ajax.METHOD_POST,
                        params: { a: 1, b: 2 }
                    }).then(function ( data ) {
                        expect(data).to.equal("content");
                        done();
                    }, function ( status ) {
                        throw new Error("unexpected status: " + status);
                    });
                });
                it("post malformed params should error", function ( done ) {
                    ajax(URL + "/post_200/", {
                        method: ajax.METHOD_POST,
                        params: { a: 1, b: 3 }
                    }).then(function ( data ) {
                        throw new Error("unexpected data: " + data);
                    }, function ( status ) {
                        expect(status).to.equal(0);
                        done();
                    });
                });
            });
            describe("catch", function ( ) {
                it("should not be called", function ( done ) {
                    ajax(URL + "/get_200/").then(done.bind(null, null),
                    function ( ) {
                        throw new Error("should not be called");
                    });
                });
            });

        });
        describe("where file does not exist", function ( ) {
            describe("then", function ( ) {
                it("should not be called", function ( done ) {
                    ajax(URL + "/get_404/").then(function ( ) {
                        throw new Error("should not be called");
                    }, done.bind(null, null));
                });
                it("should support error param", function ( done ) {
                    ajax(URL + "/get_404/").then(function ( ) {
                        throw new Error("should not be called");
                    }, done.bind(null, null));
                });
            });
            describe("catch", function ( ) {
                it("should have 404 status", function ( done ) {
                    ajax(URL + "/get_404/").catch(function ( status ) {
                        expect(status).to.equal(404);
                        done();
                    });
                });
            });
        });

    });

})( );



ajax.js

(function ( factory ) {
    "use strict";

    var old;

    if ("function" === typeof define && define.amd) {
        define(factory);
    } else if ("undefined" !== typeof module) {
        module.exports = factory();
    } else {
        old = window.ajax;
        window.ajax = factory();

        /**
         * resets the old value at ajax and returns this one
         * @memberof ajax
         * @function noConflict
         * @return {Function} ajax
         */
        window.ajax.noConflict = function ( ) {
            var tmp;
            tmp = window.ajax;
            window.ajax = old;
            return tmp;
        };
    }

})(function ( ) {
    "use strict";

    var ajax, normalize, throwable, create_xhr;

    throwable = function ( message ) {
        return new Error("ajax: " + message);
    };

    create_xhr = function ( ) {
        if (window.XMLHttpRequest) {
            return new XMLHttpRequest();
        } else if (window.ActiveXObject) {
            return new ActiveXObject("Microsoft.XMLHTTP");
        }
        throw throwable("request object not supported");
    };

    // argument normalization and sanity check
    normalize = function ( data ) {

        var result;

        // ignore extra arguments
        switch (Math.min(data.length, 2)) {
            case 1:
                if ("object" === typeof data[0]) {
                    result = data[0];
                } else if ("string" === typeof data[0]) {
                    result = { url: data[0] };
                } else {
                    result = { };
                }
                break;
            case 2:
                if ("string" === typeof data[0] &&
                    "object" === typeof data[1]) {
                    data[1].url = data[0];
                    result = data[1];
                } else {
                    result = { };
                }
                break;
            default:
                result = { };
        }

        if ("boolean" !== typeof result.auto) {
            result.auto = true;
        }

        if ("string" !== typeof result.method) {
            result.method = ajax.METHOD_GET;
        }

        if ("string" !== typeof result.type) {
            result.type = ajax.TYPE_JSON;
        }

        result.params = Object(result.params);

        if ("string" !== typeof result.url) {
            throw throwable("argument required [url] as [string]");
        }

        return result;
    };

    /**
     * responsible for setting up and monitoring an xhr request
     * @function ajax
     * @param {(string|Object)} param1 url or configuration object
     * @param {Object} [param2] configuration object
     * @param {String} [param2.url] request location
     * @param {Boolean}[param2.auto=true] send request immediately
     * @param {String} [param2.method=get] xhr method
     * @param {String} [param2.type] specifies content type
     * @param {Object} [param2.params={}] parameters to send in request
     */
    ajax = function ( /* arguments */ ) {
        var data, key, xhr, promise;

        data = normalize(arguments);
        data.post = null;

        if (data.method !== ajax.METHOD_GET) {
            if (data.type === ajax.TYPE_JSON) {
                data.post = JSON.stringify(data.params);
            } else if (data.type === ajax.TYPE_URL_ENCODED) {
                data.post = "";

                for (key in data.params) {
                    data.post += "&" + key + "=" + data.params[key];
                }

                // clean up first character
                data.post = data.post.substr(1);
            }

        } else if (Object.keys(data.params).length) {
            if (!~data.url.indexOf("?")) {
                data.url += "?";
            }

            for (key in data.params) {
                data.url += "&" + key + "=" + data.params[key];
            }
        }

        // setup xhr
        xhr = create_xhr();

        // setup promise
        promise = new window.Promise(function ( accept, reject ) {
            xhr.accept = accept;
            xhr.reject = reject;

            xhr.onreadystatechange = function ( ) {
                if (4 !== this.readyState) {
                    return void 0;
                }

                if (200 === this.status) {
                    accept(this.responseText);
                } else {
                    reject(this.status);
                }
            };
        });

        /**
         * aborts current xhr call
         * @memberof ajax#
         * @function cancel
         * @return {Object} self context for chaining
         */
        promise.cancel = function (  ) {
            xhr.abort();
            xhr.reject(ajax.STATUS_ABORT);
            return promise;
        };

        /**
         * sends the xhr
         * @function send
         * @memberof ajax#
         */
        promise.send = function (  ) {
            xhr.open(data.method, data.url, true);
            xhr.setRequestHeader("Content-Type", data.type);
            xhr.send(data.post);
            return promise;
        };

        // send the request if auto
        if (data.auto) {
            promise.send();
        }

        return promise;
    };

    /**
     * defines the xhr get request method
     * @memberof ajax
     * @member {string}
     * @static
     */
    ajax.METHOD_GET     = "GET";
    /**
     * defines the xhr post request method
     * @memberof ajax
     * @member {string}
     * @static
     */
    ajax.METHOD_POST    = "POST";
    /**
     * defines the xhr put request method
     * @memberof ajax
     * @member {string}
     * @static
     */
    ajax.METHOD_PUT     = "PUT";
    /**
     * defines the xhr update request method
     * @memberof ajax
     * @member {string}
     * @static
     */
    ajax.METHOD_UPDATE  = "UPDATE";
    /**
     * defines the xhr delete request method
     * @memberof ajax
     * @member {string}
     * @static
     */
    ajax.METHOD_DELETE  = "DELETE";

    /**
     * defines the abort status
     * @memberof ajax
     * @member {Number}
     * @static
     */
    ajax.STATUS_ABORT = -1;
    /**
     * defines the success status
     * @memberof ajax
     * @member {Number}
     * @static
     */
    ajax.STATUS_SUCCESS = 200;
    /**
     * defines the not found status
     * @memberof ajax
     * @member {Number}
     * @static
     */
    ajax.STATUS_NOT_FOUND = 404;

    /**
     * defines the url encoded type
     * @memberof ajax
     * @member {String}
     * @static
     */
    ajax.TYPE_URL_ENCODED = "application/x-www-form-urlencoded";
    /**
     * defines the json type
     * @memberof ajax
     * @member {String}
     * @static
     */
    ajax.TYPE_JSON = "application/json";

    return ajax;

});

